precompiles/xtokens/src/lib.rs

```diff
@@ -0,0 +1,417 @@
+use sp_std::{
+	boxed::Box,
+	convert::{TryFrom, TryInto},
+	marker::PhantomData,
+	vec::Vec,
+};
+use sp_weights::Weight;
+use xcm::{
+	latest::{AssetId, Fungibility, MultiAsset, MultiAssets, MultiLocation, WeightLimit},
+	VersionedMultiAsset, VersionedMultiAssets, VersionedMultiLocation,
+};
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+const DEFAULT_PROOF_SIZE: u64 = 256 * 1024;
+
+pub type XBalanceOf<Runtime> = <Runtime as orml_xtokens::Config>::Balance;
+pub type MaxAssetsForTransfer<Runtime> = <Runtime as orml_xtokens::Config>::MaxAssetsForTransfer;
+
+pub type CurrencyIdOf<Runtime> = <Runtime as orml_xtokens::Config>::CurrencyId;
+
+pub struct GetMaxAssets<R>(PhantomData<R>);
+
+impl<R> Get<u32> for GetMaxAssets<R>
+where
+	R: orml_xtokens::Config,
+{
+	fn get() -> u32 {
+		<R as orml_xtokens::Config>::MaxAssetsForTransfer::get() as u32
+	}
+}
+
+/// A precompile to wrap the functionality from xtokens
+pub struct XtokensPrecompile<Runtime>(PhantomData<Runtime>);
+
+#[precompile_utils::precompile]
+#[precompile::test_concrete_types(mock::Runtime)]
+impl<Runtime> XtokensPrecompile<Runtime>
+where
+	Runtime: orml_xtokens::Config
+		+ pallet_evm::Config
+		+ frame_system::Config
+		+ pallet_assets::pallet::Config,
+	Runtime::RuntimeCall: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	Runtime::RuntimeCall: From<orml_xtokens::Call<Runtime>>,
+	<Runtime::RuntimeCall as Dispatchable>::RuntimeOrigin: From<Option<Runtime::AccountId>>,
+	XBalanceOf<Runtime>: TryFrom<U256> + Into<U256> + solidity::Codec,
+	Runtime: EVMAddressToAssetId<AssetIdOf<Runtime>>,
+	<Runtime as orml_xtokens::Config>::CurrencyId:
+		From<<Runtime as pallet_assets::Config>::AssetId>,
+{
+	#[precompile::public("transfer(address,uint256,(uint8,bytes[]),uint64)")]
+	fn transfer(
+		handle: &mut impl PrecompileHandle,
+		currency_address: Address,
+		amount: U256,
+		destination: MultiLocation,
+		weight: u64,
+	) -> EvmResult {
+		let to_address: H160 = currency_address.into();
+
+		let asset_id = Runtime::address_to_asset_id(to_address).ok_or(
+			RevertReason::custom("Cannot convert into currency id").in_field("currencyAddress"),
+		)?;
+
+		let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
+		let amount = amount
+			.try_into()
+			.map_err(|_| RevertReason::value_is_too_large("balance type").in_field("amount"))?;
+
+		let dest_weight_limit = if weight == u64::MAX {
+			WeightLimit::Unlimited
+		} else {
+			WeightLimit::Limited(Weight::from_parts(weight, DEFAULT_PROOF_SIZE))
+		};
+
+		let call = orml_xtokens::Call::<Runtime>::transfer {
+			currency_id: asset_id.into(),
+			amount,
+			dest: Box::new(VersionedMultiLocation::V3(destination)),
+			dest_weight_limit,
+		};
+
+		RuntimeHelper::<Runtime>::try_dispatch(
+			handle,
+			Some(origin).into(),
+			call,
+			SYSTEM_ACCOUNT_SIZE,
+		)?;
+
+		Ok(())
+	}
+
+	#[precompile::public("transferWithFee(address,uint256,uint256,(uint8,bytes[]),uint64)")]
```

Comment:
I don't understand the purpose of this dual declaration, could you please explain?

Reply:
Let us discuss the naming style